src/Emulator/Extensions/Mocks/HPSHostController.cs

```diff
@@ -171,6 +174,114 @@ private byte[] GetBytesFromSlave(int count, TimeInterval timeInterval)
             return result;
         }
 
+        private void PollForError(TimeInterval timeInterval){
+            var errors = ReadCommonErrorStatusBytes(timeInterval);
+            var sw = new Stopwatch();
+            sw.Start();
+            while (errors[0] == 0x00 && sw.Elapsed < new TimeSpan(0,0,3)){
+                errors = ReadCommonErrorStatusBytes(timeInterval);
+            }
+        }
+
+        // When we need to read for more than 255 bytes, we will need this method
+        private byte[] GetLongBytes(int count, TimeInterval timeInterval) {            
+            var result = new byte[count];
+            int rounds = count /255;
+            int remainder = count % 255; 
+            for (int i = 0; i <= rounds; i ++) {
+                int toRead = i == rounds ? remainder : 255;
+                Array.Copy(GetBytesFromSlave(toRead, timeInterval), 0, result, i * 255, toRead);
+            }
+            return result;
+        }
+
+        public void StartSocketServer(TimeInterval timeInterval)
+        {  
+            Console.WriteLine(currentSlave.GetType().Name);
+            string path = "/tmp/i2c.sock";
+            if (File.Exists(path))
+            {
+                File.Delete(path);
+            }
+            var socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
+            socket.Bind(new UnixEndPoint(path));
+            socket.Listen(0);
+            Console.WriteLine("Server started waiting for client to connect...");
+            Console.WriteLine("Starting on: "+ path);
+            currentSlave.Read(0);
+            var connection = socket.Accept();
+            Console.WriteLine("Connection established");
+
+            var commandBuffer = new byte[1];
+            int length;
+            while(SocketConnected(connection)){
+                try {
+                    PollForSocketData(connection, (int)numOfBytes.MessageType, commandBuffer);
+                }
+                catch(Exception e)
+                {
+                    Console.WriteLine("{0} Exception caught.", e);
+                    continue;
+                }
+                switch(commandBuffer[0])
+                {
+                    case (byte)SocketI2CMessageType.Stop:
+                        continue;
+                    // Parse read command
+                    case (byte)SocketI2CMessageType.Read:
+                        length = GetLength(connection);
+                    
+                        byte [] dataToSend =  GetLongBytes(length, timeInterval);
+                        try 
+                        {   
+                            connection.Send(dataToSend, 0, length, SocketFlags.None);
```

Comment:
I think there's the potential for the same problem here as there was before with reading - i.e. it might not send all the bytes in one go. I haven't actually looked at the definition of the Send method, but I'm guessing that it also returns how many bytes were actually written

Reply:
It does, but I believe on blocking connections it should send it all or throw. I'd need to look into details of the unix socket implementation, but I think it's relatively safe.

Of course having a loop won't hurt

Reply:
Yeah it's a blocking kind of unix socket, so it's relatively safe, butt I guess it's probs better to have a poll too.